optuna/importance/_fanova/_tree.py

```diff
@@ -0,0 +1,313 @@
+
+    @property
+    def variance(self) -> float:
+        if self._variance is None:
+            leaf_node_indices = numpy.array(
+                [
+                    node_index
+                    for node_index in range(self._n_nodes)
+                    if self._is_node_leaf(node_index)
+                ]
+            )
+            statistics = self._statistics[leaf_node_indices]
+            values = statistics[:, 0]
+            weights = statistics[:, 1]
+            average_values = numpy.average(values, weights=weights)
+            variance = numpy.average((values - average_values) ** 2, weights=weights)
+
+            self._variance = variance
+
+        assert self._variance is not None
+        return self._variance
+
+    def get_marginal_variance(self, features: numpy.ndarray) -> float:
+        assert features.size > 0
+
+        # For each midpoint along the given dimensions, traverse this tree to compute the
+        # marginal predictions.
+        midpoints = [self._split_midpoints[f] for f in features]
+        sizes = [self._split_sizes[f] for f in features]
+
+        product_midpoints = itertools.product(*midpoints)
+        product_sizes = itertools.product(*sizes)
+
+        sample = numpy.full(self._n_features, fill_value=numpy.nan, dtype=numpy.float64)
+
+        values = []  # type: Union[List[float], numpy.ndarray]
+        weights = []  # type: Union[List[float], numpy.ndarray]
+
+        for midpoints, sizes in zip(product_midpoints, product_sizes):
+            sample[features] = numpy.array(midpoints)
+
+            value, weight = self._get_marginalized_statistics(sample)
+            weight *= numpy.prod(sizes)
+
+            values.append(value)
+            weights.append(weight)
+
+        weights = numpy.array(weights)
+        values = numpy.array(values)
+        average_values = numpy.average(values, weights=weights)
+        variance = numpy.average((values - average_values) ** 2, weights=weights)
+
+        assert variance >= 0.0
+        return variance
+
+    def _get_marginalized_statistics(self, feature_vector: numpy.ndarray) -> Tuple[float, float]:
+        assert feature_vector.size == self._n_features
+
+        marginalized_features = numpy.isnan(feature_vector)
+        active_features = ~marginalized_features
+
+        # Reduce search space cardinalities to 1 for non-active features.
+        search_spaces = self._search_spaces.copy()
+        search_spaces[marginalized_features] = [0.0, 1.0]
+
+        # Start from the root and traverse towards the leafs.
+        active_nodes = [0]
+        active_search_spaces = [search_spaces]
+
+        node_indices = []
+        active_features_cardinalities = []
+
+        while len(active_nodes) > 0:
+            node_index = active_nodes.pop()
+            search_spaces = active_search_spaces.pop()
+
+            feature = self._get_node_split_feature(node_index)
+            if feature >= 0:  # Not leaf.
```

Comment:
nit: Using `_is_node_leaf` will slightly improve readability.
```suggestion
            if not self._is_node_leaf(node_index):
                feature = self._get_node_split_feature(node_index)
```

Reply:
That's a perfectly sensible comment. This is actually in favor of performance. It contributes to around 10% of the gain in https://github.com/optuna/optuna/pull/1106#issuecomment-619388187.

Reply:
Added comments about intention https://github.com/optuna/optuna/pull/1106/commits/19fc7b4c834121416f84d5d1537a5e4f9583cd8f. What do you think?